File: python/neuralsemantic/vector/vector_store.py
"""ChromaDB-based vector storage for semantic similarity."""

import logging
import numpy as np
from pathlib import Path
from typing import List, Dict, Any, Optional

try:
    import chromadb
    from chromadb.config import Settings
    CHROMADB_AVAILABLE = True
except ImportError:
    CHROMADB_AVAILABLE = False
    chromadb = None

from ..core.types import Pattern, SimilarPattern
from ..core.exceptions import VectorStoreError
from .embeddings import EmbeddingManager

logger = logging.getLogger(__name__)


class VectorStore:
    """ChromaDB-based vector storage for semantic similarity."""

    def __init__(self, config):
        if not CHROMADB_AVAILABLE:
            logger.warning("ChromaDB not available, vector features disabled")
            self.enabled = False
            return
            
        self.config = config
        self.enabled = True
        self._initialize_client()
        self.embedding_manager = EmbeddingManager(config.model_name)

    def _initialize_client(self) -> None:
        """Initialize ChromaDB client."""
        try:
            # Ensure directory exists
            persist_dir = Path(self.config.persist_directory)
            persist_dir.mkdir(parents=True, exist_ok=True)
            
            # Initialize client
            self.client = chromadb.PersistentClient(
                path=str(persist_dir),
                settings=Settings(
                    anonymized_telemetry=False,
                    allow_reset=True
                )
            )
            
            # Get or create collection
            try:
                # Try to get existing collection first
                self.collection = self.client.get_collection(name="compression_patterns")
            except:
                # Create new collection with proper metadata format
                self.collection = self.client.create_collection(
                    name="compression_patterns",
                    metadata={"hnsw_space": "cosine"}  # Simplified metadata
                )
            
            logger.info(f"Vector store initialized at {self.config.persist_directory}")
            
        except Exception as e:
            logger.error(f"Failed to initialize vector store: {e}")
            self.enabled = False
            raise VectorStoreError(f"Vector store initialization failed: {e}")

    def add_pattern(self, pattern: Pattern) -> None:
        """Add pattern to vector store with metadata."""
        if not self.enabled:
            return
            
        try:
            # Generate embedding
            embedding = self.embedding_manager.get_embedding(pattern.original)
            
            # Prepare metadata
            metadata = {
                "pattern_id": pattern.id or 0,
                "compressed": pattern.compressed,
                "pattern_type": pattern.pattern_type.value,
                "domain": pattern.domain,
                "priority": pattern.priority,
                "compression_ratio": len(pattern.compressed) / len(pattern.original),
                "language": pattern.language
            }

            # Add to collection
            self.collection.add(
                embeddings=[embedding.tolist()],
                documents=[pattern.original],
                metadatas=[metadata],
                ids=[f"pattern_{pattern.id or hash(pattern.original)}"]
            )

        except Exception as e:
            logger.error(f"Failed to add pattern to vector store: {e}")

    def find_similar_patterns(self, text: str, n_results: int = 5,
                            threshold: float = 0.8, domain: Optional[str] = None) -> List[SimilarPattern]:
        """Find semantically similar patterns."""
        if not self.enabled:
            return []
            
        try:
            # Generate query embedding
            query_embedding = self.embedding_manager.get_embedding(text)

            # Prepare where clause for filtering
            where_clause = {}
            if domain:
                where_clause["domain"] = domain

            # Query vector store
            results = self.collection.query(
                query_embeddings=[query_embedding.tolist()],
                n_results=n_results * 2,  # Get more to filter
                where=where_clause if where_clause else None,
                include=["metadatas", "documents", "distances"]
            )

            # Process results
            similar_patterns = []
            if results['distances'] and results['distances'][0]:
                for i, distance in enumerate(results['distances'][0]):
                    similarity = 1 - distance

                    if similarity >= threshold:
                        metadata = results['metadatas'][0][i]
                        similar_patterns.append(SimilarPattern(
                            original=results['documents'][0][i],
                            compressed=metadata['compressed'],
                            similarity=similarity,
                            pattern_type=metadata['pattern_type'],
                            domain=metadata['domain'],
                            priority=metadata['priority'],
                            confidence=similarity * metadata['priority'] / 1000
                        ))

            # Sort by similarity and return top results
            similar_patterns.sort(key=lambda x: x.similarity, reverse=True)
            return similar_patterns[:n_results]

        except Exception as e:
            logger.error(f"Vector similarity search failed: {e}")
            return []

    def bulk_add_patterns(self, patterns: List[Pattern]) -> None:
        """Efficiently add multiple patterns."""
        if not self.enabled or not patterns:
            return

        try:
            embeddings = []
            documents = []
            metadatas = []
            ids = []

            for pattern in patterns:
                embedding = self.embedding_manager.get_embedding(pattern.original)
                embeddings.append(embedding.tolist())
                documents.append(pattern.original)
                metadatas.append({
                    "pattern_id": pattern.id or 0,
                    "compressed": pattern.compressed,
                    "pattern_type": pattern.pattern_type.value,
                    "domain": pattern.domain,
                    "priority": pattern.priority,
                    "compression_ratio": len(pattern.compressed) / len(pattern.original),
                    "language": pattern.language
                })
                ids.append(f"pattern_{pattern.id or hash(pattern.original)}")

            # Batch add to collection
            self.collection.add(
                embeddings=embeddings,
                documents=documents,
                metadatas=metadatas,
                ids=ids
            )

            logger.info(f"Added {len(patterns)} patterns to vector store")

        except Exception as e:
            logger.error(f"Bulk add to vector store failed: {e}")

    def get_collection_stats(self) -> Dict[str, Any]:
        """Get vector store statistics."""
        if not self.enabled:
            return {"enabled": False}
            
        try:
            count = self.collection.count()
            return {
                "enabled": True,
                "total_patterns": count,
                "collection_name": self.collection.name,
                "model_name": self.embedding_manager.model_name
            }
        except Exception as e:
            logger.error(f"Failed to get collection stats: {e}")
            return {"enabled": True, "error": str(e)}

    def clear_collection(self) -> None:
        """Clear all patterns from vector store."""
        if not self.enabled:
            return
            
        try:
            # Delete and recreate collection
            self.client.delete_collection("compression_patterns")
            self.collection = self.client.get_or_create_collection(
                name="compression_patterns",
                metadata={
                    "hnsw:space": "cosine",
                    "hnsw:M": 16,
                    "hnsw:ef_construction": 200
                }
            )
            logger.info("Vector store collection cleared")
        except Exception as e:
            logger.error(f"Failed to clear collection: {e}")

    def search_by_metadata(self, metadata_filter: Dict[str, Any], 
                          n_results: int = 20) -> List[Dict[str, Any]]:
        """Search patterns by metadata."""
        if not self.enabled:
            return []
            
        try:
            results = self.collection.get(
                where=metadata_filter,
                limit=n_results,
                include=["metadatas", "documents"]
            )
            
            patterns = []
            if results['documents']:
                for i, doc in enumerate(results['documents']):
                    patterns.append({
                        "original": doc,
                        "metadata": results['metadatas'][i]
                    })
            
            return patterns
            
        except Exception as e:
            logger.error(f"Metadata search failed: {e}")
            return []

    def update_pattern_metadata(self, pattern_id: str, metadata: Dict[str, Any]) -> bool:
        """Update pattern metadata."""
        if not self.enabled:
            return False
            
        try:
            self.collection.update(
                ids=[pattern_id],
                metadatas=[metadata]
            )
            return True
        except Exception as e:
            logger.error(f"Failed to update pattern metadata: {e}")
            return False

    def delete_pattern(self, pattern_id: str) -> bool:
        """Delete pattern from vector store."""
        if not self.enabled:
            return True  # No-op if disabled
            
        try:
            self.collection.delete(ids=[pattern_id])
            return True
        except Exception as e:
            logger.error(f"Failed to delete pattern from vector store: {e}")
            return False

File: packages/core/src/vector/vector-store.ts
/**
 * Vector Store for Neural Semantic Compiler
 */

import {
  VectorStore as IVectorStore,
  VectorConfig,
  Pattern,
  SimilarPattern
} from '@neurosemantic/types';

export class VectorStore implements IVectorStore {
  public enabled: boolean = false;
  private config: VectorConfig;
  private chromaClient: any;
  private collection: any;
  private embeddingManager: any;

  constructor(config: VectorConfig) {
    this.config = config;
  }

  async initialize(): Promise<void> {
    try {
      // Try to initialize ChromaDB
      const { ChromaClient } = await import('chromadb');
      
      this.chromaClient = new ChromaClient({
        path: this.config.persistDirectory
      });

      // Get or create collection
      try {
        this.collection = await this.chromaClient.getCollection({
          name: 'neural_semantic_patterns'
        });
      } catch (error) {
        this.collection = await this.chromaClient.createCollection({
          name: 'neural_semantic_patterns',
          metadata: { description: 'Neural Semantic Compiler patterns' }
        });
      }

      this.enabled = true;
      console.log('Vector store initialized successfully');

    } catch (error) {
      console.warn('Vector store not available:', error);
      this.enabled = false;
    }
  }

  async addPattern(pattern: Pattern): Promise<void> {
    if (!this.enabled || !this.collection) {
      return;
    }

    try {
      await this.collection.add({
        ids: [`pattern_${pattern.id}`],
        documents: [pattern.original],
        metadatas: [{
          pattern_id: pattern.id,
          compressed: pattern.compressed,
          pattern_type: pattern.patternType,
          domain: pattern.domain,
          priority: pattern.priority,
          success_rate: pattern.successRate
        }]
      });
    } catch (error) {
      console.error('Failed to add pattern to vector store:', error);
    }
  }

  async findSimilarPatterns(
    text: string,
    nResults: number = 10,
    threshold: number = 0.7,
    domain?: string
  ): Promise<SimilarPattern[]> {
    if (!this.enabled || !this.collection) {
      return [];
    }

    try {
      const queryFilter = domain ? { domain } : undefined;

      const results = await this.collection.query({
        queryTexts: [text],
        nResults,
        where: queryFilter
      });

      const similarPatterns: SimilarPattern[] = [];

      if (results.documents && results.metadatas && results.distances) {
        for (let i = 0; i < results.documents[0].length; i++) {
          const distance = results.distances[0][i];
          const similarity = 1 - distance; // Convert distance to similarity

          if (similarity >= threshold) {
            const metadata = results.metadatas[0][i];
            
            similarPatterns.push({
              original: results.documents[0][i],
              compressed: metadata.compressed,
              similarity,
              patternType: metadata.pattern_type,
              domain: metadata.domain,
              priority: metadata.priority,
              confidence: similarity * (metadata.success_rate || 0.5)
            });
          }
        }
      }

      return similarPatterns.sort((a, b) => b.confidence - a.confidence);

    } catch (error) {
      console.error('Failed to find similar patterns:', error);
      return [];
    }
  }

  async bulkAddPatterns(patterns: Pattern[]): Promise<void> {
    if (!this.enabled || !this.collection) {
      return;
    }

    try {
      const ids = patterns.map(p => `pattern_${p.id}`);
      const documents = patterns.map(p => p.original);
      const metadatas = patterns.map(p => ({
        pattern_id: p.id,
        compressed: p.compressed,
        pattern_type: p.patternType,
        domain: p.domain,
        priority: p.priority,
        success_rate: p.successRate
      }));

      await this.collection.add({
        ids,
        documents,
        metadatas
      });

    } catch (error) {
      console.error('Failed to bulk add patterns:', error);
    }
  }

  async getCollectionStats(): Promise<Record<string, any>> {
    if (!this.enabled || !this.collection) {
      return {
        enabled: false,
        total_patterns: 0
      };
    }

    try {
      const count = await this.collection.count();
      
      return {
        enabled: true,
        total_patterns: count,
        collection_name: 'neural_semantic_patterns',
        config: {
          model_name: this.config.modelName,
          similarity_threshold: this.config.similarityThreshold,
          max_results: this.config.maxResults
        }
      };
    } catch (error) {
      console.error('Failed to get collection stats:', error);
      return {
        enabled: false,
        error: error instanceof Error ? error.message : String(error)
      };
    }
  }

  async clearCollection(): Promise<void> {
    if (!this.enabled || !this.chromaClient) {
      return;
    }

    try {
      await this.chromaClient.deleteCollection({
        name: 'neural_semantic_patterns'
      });

      // Recreate empty collection
      this.collection = await this.chromaClient.createCollection({
        name: 'neural_semantic_patterns',
        metadata: { description: 'Neural Semantic Compiler patterns' }
      });

    } catch (error) {
      console.error('Failed to clear collection:', error);
    }
  }
}